### ISSUE TITLE
End-to-End Memory Network

### ISSUE THREAD

**Issue by yskim0** [yskim0/One-Day-One-Paper#47]
- abstract
    - recurrent attention model over a possibly large external memory
    - Memory Network에서 따왔지만 end-to-end training이 가능하고 이로 인해 학습할 때 less supervision, 현실적인 셋팅에서 적용가능해진다는 점에서 차별점이 있다.
    - output(answer)을 내기 위해 muiltiple computational steps(`hops`) 를 사용한다.
        - 그냥 말그대로 여러 단계를 거침.
- introduction
    - we present a novel recurrent neural network (RNN) architecture where the recurrence reads from a possibly large external memory multiple times before outputting a symbol
        - **continuous form of the Memory Network**
            - 앞선 연구들(memory networks, neural machine translation by jointly learning to align and translate, neural turing machines...)은 continuous represetnation을 저장하는 스토리지를 따로 사용함 → reads from, write to the stroage
                - 이전까지는 완전한 continuous가 아니라는 것을 말하고 싶어하는 듯하다
    - 이전 메모리 네트워크는 backpropagation을 통한 학습이 어려웠다. ← 매 레이어마다 supervision이 필요했기 때문에라고 한다.
        - Why?
    - multiple computational steps("hops") per output symbol
        - output symbol?
- approach
    - input $x_1, ... , x_n$, a query $q$, an answer $a$
    - $x_i, q, a$ contains symbols coming from a dictionary with V words
    - input x와 query의 continuous representation을 찾고, 이걸 multiple hops를 통해 처리함으로써 output a를 만든다.
        - backpropagation을 가능하게 함
        - through multiple memory accesses back to the input during training.

    ![https://s3-us-west-2.amazonaws.com/secure.notion-static.com/87321646-f0bc-4a86-82ef-9db3f0e83fbe/Untitled.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/87321646-f0bc-4a86-82ef-9db3f0e83fbe/Untitled.png)

    - (a) **single layer**

        ![https://s3-us-west-2.amazonaws.com/secure.notion-static.com/9cf28538-4754-4d0f-9b35-677faeb42f4a/Untitled.png](https://s3-us-west-2.amazonaws.com/secure.notion-static.com/9cf28538-4754-4d0f-9b35-677faeb42f4a/Untitled.png)

        - single memory hop operation  ⇒ 해당 레이어를 stack하면 전체 아키텍처가 됨
            - input memory representation
                - input set $x_1, ..., x_i$(즉 $\left \{x_i \right\}$)는 d-dimension의 memory vector $\left \{m_i \right \}$로 변환됨
                (임베딩 벡터로 변환된 후 메모리에 저장된다.)
                    - HOW? → embedding matrix `A` 를 사용 ($A \in \mathbb{R}^{d\times V}$)
                - query q 또한 embedding matrix `B` 를 사용하여 임베딩됨. → internal state $u$ 를 얻기 위해 임베딩시킴
                    - 이때 B는 A와 같은 dimension을 가짐
                    - ~~internal state 의 역할이 뭘까? ⇒ 해당 쿼리를 represent하는 벡터~~
                - matching between $u$ and each memory $m_i$ by taking inner product follwed by a softmax

                    $$p_i = Softmax(u^Tm_i)$$

                    - 내적 → 현재 (internal) 스테이트와 지금까지 써진 메모리 셀들과의 연관성 내지는 유사도를 구하는 걸로 이해함. Softmax를 취했으니까 비슷한 스테이트/셀끼리는 높은 값을 지닐 것임. 확률로 이해
            - Output Memory Representation
                - each $x_i$ has a corresponding output vector $c_i$(embedding matrix `C` 를 사용하여 임베딩됨)
                    - output vector c의 의미가 완벽히 이해되지는 않음
                - 아웃풋(response vector) $o$

                    $$o = \sum_ip_ic_i$$

                    이 $o$ 은 smooth해져서 gradient 계산과 backprop.이 쉽게 계산됨.

            - Generating the final prediction
                - In the single layer, the sum of the output vector $o$ and the input embedding $u$는 final weight matrix $W \in \mathbb{R}^{V\times d}$를 지나고 소프트맥스를 지나 predicted label을 만들어냄

                $$\hat{a} = Softmax(W(o+u))$$

            - 학습 중에 `A, B, C and W` 는 $\hat{a}$, $a$ 사이의 크로스엔트로피 로스를 최소화하는 방향으로 학습된다. SGD를 사용해서 업데이트
        - Multiple layers
            - single layer를 K hop stack한다고 볼 수 있다.
            - 첫번째 레이어를 제외한 레이어의 q vector representation은 이전 레이어의 아웃풋 벡터인 o^k와 query vector u^k를 더해서 사용한다.

            $$u^{k+1} = u^k + o^k$$

            - 각 레이어는 x_i를 임베딩하기 위해 개별적인 $A^k, C^k$가 존재한다.  하지만 공유될 수도 있다~

                > **Weight-tying** is where you have a language model and **use the same weight matrix for the input-to-embedding layer (the input embedding) and the hidden-to-softmax layer (the output embedding).**

                1. Adjacent 방식 ; 인접한 임베딩 매트릭스를 share한다
                    - $A^{k+1} = C^k$
                    - $W^T = C^K$ : prediction을 위한 matrix W는 마지막 layer의 C matrix를 transpose한 것과 같다.
                    - $B = A^1$ : 쿼리를 임베딩하는 B는 첫 레이어의 A 매트릭스와 동일하다.
                2. Layer-wise (RNN-like) : 인풋 임베딩과 아웃풋 임베딩을 모두 같은 매트릭스를 쓰게 함.
                    - $A^1=A^2=...=A^K$, $C^1=C^2=...=C^K$
                    - linear mapping function H를 사용 (we found it useful to add a linear mapping H to the update of $u$ between hops)
                    $u^{k+1} = Hu^k + o^k$
            - 마지막 레이어에서의 prediction은

            $$\hat{a} = Softmax(Wu^{K+1}) = Softmax(W(o^K+u^K))$$

**Comment by yskim0** [comment]
사진이랑 수식이 다 깨지네요; ;ㅠ
깔끔한 건 여기서 볼 수 있습니다... 
https://beautiful-beetle-574.notion.site/End-to-End-Memory-Network-1a43fafc22774bce97ded4cab7e6f4d1
